crates/p2p/src/main_loop.rs

```diff
@@ -176,11 +176,23 @@ impl MainLoop {
             // Connection management
             // ===========================
             SwarmEvent::ConnectionEstablished {
-                peer_id, endpoint, ..
+                peer_id,
+                endpoint,
+                num_established,
+                ..
             } => {
-                // If this is an incoming connection, we have to prevent the peer from
-                // reconnecting too quickly.
                 if endpoint.is_listener() {
+                    // This is an incoming connection.
+
+                    // Only allow one connection per peer.
+                    if num_established.get() > 1 {
+                        tracing::warn!(%peer_id, "Peer has more than one connection, closing");
```

Comment:
Should be much lower level since this can be trivially caused by external parties. Debug or even trace 😅